### PR TITLE
Python version

### PR DESCRIPTION
**Pull request by terakilobyte**
Also please review @ShaneHarvey

### CODE REVIEW

File: python/clients.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        "local": {
+            "key": local_master_key,
+        },
+    }
+
+    client_prototype = ClientBuilder(kms_providers=kms_providers)
+
+    # make sure you paste in your base64 key that was generated from
+    # make-local-data-key.py
+    data_key = "jXjXI3pAR+Cbd7rJufnQKA=="  # replace with your base64 data key!
```
Comment (terakilobyte) [inline python/clients.py]: I need to change this to None

File: python/clients.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+from helpers import read_master_key, ClientBuilder, create_json_schema
+
+"""
+This file is meant to be run to demonstrate CSFLE in action. Prior to running
+this file, ensure you've run make-local-data-key.py
```
Comment (ccho-mongodb) [inline python/clients.py]: I think we should use PEP8 recommendations for module naming for python scripts: https://www.python.org/dev/peps/pep-0008/#package-and-module-names
tl;dr prefer single-word file names and otherwise use underscores

File: python/helpers.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        self.key_db = key_db
+        self.key_coll = key_coll
+        self.key_vault_namespace = f"{key_db}.{key_coll}"
+        self.schema = schema
+        self.connection_string = connection_string
+        self.mongocryptd_bypass_spawn = mongocryptd_bypass_spawn
+        self.mongocryptd_spawn_path = mongocryptd_spawn_path
+
+        fle_opts = AutoEncryptionOpts(
+            self.kms_providers,
```
Comment (ccho-mongodb) [inline python/helpers.py]: This line is not consistent with line 89 in which you're using the values of the params vs the instance fields. Would it be helpful for the user to stay consistent?

File: python/helpers.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            self.key_vault_namespace,
+            mongocryptd_bypass_spawn=self.mongocryptd_bypass_spawn,
+            mongocryptd_spawn_path=self.mongocryptd_spawn_path)
+
+        self.encrypted_client = MongoClient(
+            self.connection_string, auto_encryption_opts=fle_opts)
+
+        self.regular_client = MongoClient(self.connection_string)
+
+        self.key_vault = self.encrypted_client.get_database(
```
Comment (ccho-mongodb) [inline python/helpers.py]: Was this linebreak intentional and if so is it a PEP8 recommendation? I would have expected it to be a linebreak between chained functions instead.

File: python/helpers.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                                    unique=True,
+                                    partialFilterExpression={
+                                        "keyAltNames": {
+                                            "$exists": True
+                                        }
+                                    })
+
+    def find_or_create_data_key(self):
+        """
+        In the guide, we create the data key and then show that it is created by
```
Comment (ccho-mongodb) [inline python/helpers.py]: Would it be helpful to provide a url back to the CSFLE use case guide here?

File: python/helpers.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        else:
+            uuid_data_key_id = data_key["_id"]
+
+        base_64_data_key_id = base64.b64encode(
+            uuid_data_key_id.bytes).decode("utf-8")
+
+        print("Base64 data key Id, paste this into clients.py: ",
+              base_64_data_key_id)
+
+        return uuid_data_key_id
```
Comment (ccho-mongodb) [inline python/helpers.py]: I wasn't able to find any usage of this return value (`uuid_data_key_id`), is it needed?

File: python/helpers.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    def get_csfle_enabled_client(self, schema=None):
+        if schema is None:
+            raise ValueError("schema is required")
+        fle_opts = AutoEncryptionOpts(
+            self.kms_providers,
+            self.key_vault_namespace,
+            mongocryptd_bypass_spawn=self.mongocryptd_bypass_spawn,
+            mongocryptd_spawn_path=self.mongocryptd_spawn_path,
+            schema_map=schema)
+        self.encrypted_client.close()
```
Comment (ccho-mongodb) [inline python/helpers.py]: Is this needed here? Shouldn't any close be in the same block in which a client is used?

File: python/helpers.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                    "local", key_alt_names=["demo-data-key"])
+                uuid_data_key_id = UUID(bytes=data_key)
+
+        else:
+            uuid_data_key_id = data_key["_id"]
+
+        base_64_data_key_id = base64.b64encode(
+            uuid_data_key_id.bytes).decode("utf-8")
+
+        print("Base64 data key Id, paste this into clients.py: ",
```
Comment (ccho-mongodb) [inline python/helpers.py]: Could this be more reusable if the function just returned the value and the calling functions determine what to do with it (e.g. print it)?

File: python/helpers.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+        if data_key is None:
+            with ClientEncryption(self.kms_providers,
+                                  self.key_vault_namespace,
+                                  self.encrypted_client,
+                                  CodecOptions(uuid_representation=STANDARD)
+                                  ) as client_encryption:
+
+                data_key = client_encryption.create_data_key(
+                    "local", key_alt_names=["demo-data-key"])
```
Comment (ccho-mongodb) [inline python/helpers.py]: Should you use `self.key_alt_name` here instead of specifying the value a second time?

File: python/helpers.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+        self.encrypted_client = MongoClient(
+            self.connection_string, auto_encryption_opts=fle_opts)
+
+        self.regular_client = MongoClient(self.connection_string)
+
+        self.key_vault = self.encrypted_client.get_database(
+            self.key_db).get_collection(self.key_coll)
+
+        # clients are required to create a unique sparse index on keyAltNames
```
Comment (ccho-mongodb) [inline python/helpers.py]: Is sparse a requirement? The `getKeyVault()` shell method did not specify sparse, and I think you are not specifying it either.

File: python/README.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+1. Start a locally running `mongod` version >= 4.2 running on port 27017
+2. It is recommended you create a local environment to encapsulate the
+   dependencies. We use **pyenv**
+
+   ```python
+   pyenv virtualenv 3.7.3 csfle
+   pyenv local csfle
+   pip install -r requirements.txt
+   ```
+
```
Comment (ccho-mongodb) [inline python/README.md]: Would it be helpful to have them locate and specify the path for mongocryptd on their system?

File: python/helpers.py (hunk 4 above)
Comment (terakilobyte) [reply python/helpers.py]: yes, the linebreak is intentional and formatted via `autopep8`

**Comment by terakilobyte** [comment]
> I think we should use PEP8 recommendations for module naming for python scripts: https://www.python.org/dev/peps/pep-0008/#package-and-module-names
tl;dr prefer single-word file names and otherwise use underscores 

Ok

File: python/README.md (hunk 11 above)
Comment (terakilobyte) [reply python/README.md]: addressed with usage string

File: python/clients.py (hunk 2 above)
Comment (terakilobyte) [reply python/clients.py]: Good catch, done!

**Comment by terakilobyte** [comment]
@ccho-mongodb nits addressed!

File: python/README.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
+# Python CSFLE Example
+
+## Steps
+
+1. Start a locally running `mongod` version >= 4.2 running on port 27017
+2. It is recommended you create a local environment to encapsulate the
```
Comment (ccho-mongodb) [inline python/README.md]: Nitpick:
It is recommended you -> We recommend that you

File: python/README.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+   dependencies. We use **pyenv**
```
Comment (ccho-mongodb) [inline python/README.md]: Nitpick: a link may be useful since `pyenv` is not part of a python base install.

File: python/clients.py (hunk 2 above)
Comment (ccho-mongodb) [inline python/clients.py]: Looks like this instance still contains the old name.

File: python/README.md (hunk 11 above)
Comment (ccho-mongodb) [inline python/README.md]: May be useful to also mention what the `master-key.txt` file contains.

File: python/clients.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        }
+    }
+
+    regular_client_patients_coll = regular_client.get_database(
+        "medicalRecords").get_collection("patients")
+
+    csfle_client_patients_coll = csfle_enabled_client.get_database(
+        "medicalRecords").get_collection("patients")
+
+    # performing the insert operation with the csfle enabled client
```
Comment (ccho-mongodb) [inline python/clients.py]: Nitpick: did you mean "perform the insert"

Comment (ccho-mongodb) [reply python/clients.py]: From reading your following code comments, it looks like you meant to use the present participle which implies present continuous tense. I usually comment using either an imperative or simple present tense, e.g.:
imperative - "Perform the insert"
simple present - "Performs the insert"

File: python/helpers.py [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    def __init__(self,
+                 kms_providers=None,
+                 key_alt_name="demo-data-key",
+                 key_db="encryption",
+                 key_coll="__keyVault",
+                 schema=None,
+                 connection_string="mongodb://localhost:27017",
+                 mongocryptd_bypass_spawn=False,
+                 mongocryptd_spawn_path="mongocryptd"):
+        """
```
Comment (ccho-mongodb) [inline python/helpers.py]: Nitpick: this comment block should probably be located above the class declaration rather than within the `__init__` function

File: python/helpers.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+                               .decode("utf-8"))
+
+        return base_64_data_key_id
+
+    def get_regular_client(self):
+        return MongoClient(self.connection_string)
+
+    def get_csfle_enabled_client(self, schema=None):
+        if schema is None:
+            raise ValueError("schema is required")
```
Comment (ShaneHarvey) [inline python/helpers.py]: If schema is required can we just make it a required argument? `, schema=None)` -> `, schema)`

File: python/helpers.py [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+                self.kms_providers,
+                self.key_vault_namespace,
+                mongocryptd_bypass_spawn=self.mongocryptd_bypass_spawn,
+                mongocryptd_spawn_path=self.mongocryptd_spawn_path,
+                schema_map=schema)
+        )
+
+    def create_json_schema(self, data_key=None):
+        if data_key is None:
+            raise ValueError("data_key must be provided")
```
Comment (ShaneHarvey) [inline python/helpers.py]: If data_key is required can we just make it a required argument? , `data_key =None)` -> `, data_key)`

File: python/clients.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+        "local": {
+            "key": local_master_key,
+        },
+    }
+
+    client_builder = CsfleHelper(kms_providers=kms_providers)
+
+    # make sure you paste in your base64 key that was generated from
+    # make_data_key.py
+    data_key = None  # replace with your base64 data key!
```
[the remark this answers is not in the document]
Comment (terakilobyte) [reply python/clients.py]: We debated on this for a bit, and decided we wanted them to have to edit the file just a bit to attempt to foster some learning.

File: python/clients.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+            "provider": "MaestCare",
+            "policyNumber": 123142
+        }
+    }
+
+    regular_client_patients_coll = regular_client.get_database(
+        "medicalRecords").get_collection("patients")
+
+    csfle_client_patients_coll = csfle_enabled_client.get_database(
+        "medicalRecords").get_collection("patients")
```
[the remark this answers is not in the document]
Comment (terakilobyte) [reply python/clients.py]: There's no reason, other than they were there!

File: python/README.md [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+4. Run the `make_data_key.py` script to make a data key. If there is an
+   existing data key in the **encryption.__keyVault** collection this script
+   will not create a duplicate data key.
+
+   ```python
+   python make_data_key.py
+   ```
+
+5. Run the `client.py` script to insert a document with the CSFLE-enabled client
```
[the remark this answers is not in the document]
Comment (terakilobyte) [reply python/README.md]: Good suggestion. Renamed to `app.py`

File: python/helpers.py [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+                 key_alt_name="demo-data-key",
+                 key_db="encryption",
+                 key_coll="__keyVault",
+                 schema=None,
+                 connection_string="mongodb://localhost:27017",
+                 mongocryptd_bypass_spawn=False,
+                 mongocryptd_spawn_path="mongocryptd"):
+        """If mongocryptd
+        is not installed to in your search path, ensure you override
+        mongocryptd_spawn_path
```
[the remark this answers is not in the document]
Comment (terakilobyte) [reply python/helpers.py]: I added a comment above to clarify this

File: python/helpers.py [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+
+"""Helpers for CSFLE implementation."""
+
+import base64
+
+from pymongo import MongoClient
+from pymongo.encryption_options import AutoEncryptionOpts
+from pymongo.encryption import ClientEncryption
+from bson.codec_options import CodecOptions
+from bson.binary import Binary, STANDARD, UUID
```
Comment (ShaneHarvey) [inline python/helpers.py]: We should import UUID from the standard library like this:
```
from uuid import UUID
```

File: python/helpers.py [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+                mongocryptd_spawn_path=self.mongocryptd_spawn_path,
+                schema_map=schema)
+        )
+
+    def create_json_schema(self, data_key):
+        return {
+            "medicalRecords.patients": {
+                "bsonType": "object",
+                "encryptMetadata": {
+                    "keyId": [Binary(base64.b64decode(data_key), 4)]
```
Comment (ShaneHarvey) [inline python/helpers.py]: Suggest using UUID_SUBTYPE instead of 4:
```
from bson.binary import UUID_SUBTYPE
```

File: python/helpers.py [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+
+from pymongo import MongoClient
+from pymongo.encryption_options import AutoEncryptionOpts
+from pymongo.encryption import ClientEncryption
+from bson.codec_options import CodecOptions
+from bson.binary import Binary, STANDARD, UUID
+
+
+def read_master_key(path="./master-key.txt"):
+    return Binary(open(path, "rb").read(96))
```
Comment (ShaneHarvey) [inline python/helpers.py]: This should use a with statement so that the file is automatically closed:
```
with open(path, "rb") as f:
    return f.read(96)
```
